### PR TITLE
Basic DataFusion Select Functionality

### PR DESCRIPTION
**Pull request by jdye64**
Basic SELECT functionality working, PyTests that are not working due to other issues are referenced in the `@pytest.mark.skip(...)`

This closes: #468 
This closes: #490 
This closes: #496 
This closes: #483

### CODE REVIEW

**Comment by jdye64** [comment]
It will be much easier to review this PR after [471](https://github.com/dask-contrib/dask-sql/pull/471) is merged. Feel free to wait until after that for a review.

File: dask_planner/src/sql/types.rs [hunk 1]
```diff
@@ -1,107 +1,312 @@
...
+                            }
+                            // Default to Nanosecond which is common if not present
+                            None => TimeUnit::Nanosecond,
+                        };
+                        (unit, tz)
+                    }
+                    // Default to Nanosecond and None for tz which is common if not present
+                    None => (TimeUnit::Nanosecond, None),
+                };
+                DataType::Timestamp(unit, tz)
```
Comment (ayushdg) [inline dask_planner/src/sql/types.rs:116]: Do we want to pass the same timezone even when the type is a `timestamp` type without a timezone?

File: dask_sql/mappings.py [hunk 2]
```diff
@@ -128,10 +123,10 @@ def sql_to_python_value(sql_type: str, literal_value: Any) -> Any:
...
 
-    elif sql_type.startswith("INTERVAL"):
+    elif sql_type == SqlTypeName.INTERVAL:
         # check for finer granular interval types, e.g., INTERVAL MONTH, INTERVAL YEAR
         try:
-            interval_type = sql_type.split()[1].lower()
+            interval_type = str(sql_type).split()[1].lower()
 
             if interval_type in {"year", "quarter", "month"}:
                 # if sql_type is INTERVAL YEAR, Calcite will covert to months
```
Comment (ayushdg) [inline dask_sql/mappings.py:132]: Some of these earlier assumptions about calcite may not hold true for datafusion and will need updating as and when mappings are encountered

File: dask_sql/mappings.py [hunk 3]
```diff
@@ -270,9 +261,7 @@ def cast_column_type(
     """
     current_type = df[column_name].dtype
 
-    logger.debug(
-        f"Column {column_name} has type {current_type}, expecting {expected_type}..."
-    )
+    print(f"Column {column_name} has type {current_type}, expecting {expected_type}...")
```
Comment (ayushdg) [inline dask_sql/mappings.py:264]: Debug print

File: dask_sql/physical/rel/logical/project.py [hunk 4]
```diff
@@ -78,7 +81,10 @@ def convert(self, rel: "LogicalPlan", context: "dask_sql.Context") -> DataContai
         # Make sure the order is correct
         cc = cc.limit_to(column_names)
 
-        cc = self.fix_column_to_row_type(cc, column_names)
+        cc = self.fix_column_to_row_type(cc, rel.getRowType())
         dc = DataContainer(df, cc)
-        dc = self.fix_dtype_to_row_type(dc, rel.table())
+        dc = self.fix_dtype_to_row_type(dc, rel.getRowType())
+
+        print(f"After Project: {dc.df.head(10)}")
```
Comment (ayushdg) [inline dask_sql/physical/rel/logical/project.py:88]: Same as above

File: dask_sql/mappings.py (hunk 2 above)
Comment (jdye64) [reply dask_sql/mappings.py:132]: Completely agree. Even though all the current tests are passing I am sure there are some missing edge cases. I was thinking about opening a issue just for that alone and was hoping that you might be able to help with that one since you have done a lot of work in the date/time area. Could I open an issue and assign it to you for a subsequent PR?

File: dask_planner/src/sql/types.rs (hunk 1 above)
Comment (jdye64) [reply dask_planner/src/sql/types.rs:116]: I honestly don't know? You would know much better. Which way is the correct way?

Comment (ayushdg) [reply dask_planner/src/sql/types.rs:116]: A default timestamp (without timezone) based on postresql is interpreted as the timezone in utc irrespective of the offset provided. So in this case the tz should always be utc.

Comment (jdye64) [reply dask_planner/src/sql/types.rs]: That might have been handled by the library but either way I added a check just to be sure.

File: continuous_integration/recipe/meta.yaml [hunk 5]
```diff
@@ -30,6 +30,7 @@ requirements:
     - setuptools-rust>=1.1.2
   run:
     - python
+    - setuptools-rust>=1.1.2
```
Comment (ayushdg) [inline continuous_integration/recipe/meta.yaml:33]: Based on the discussion [here](https://github.com/dask-contrib/dask-sql/pull/471#discussion_r853217325) is this needed?

File: continuous_integration/recipe/run_test.py [hunk 6]
```diff
@@ -13,19 +13,21 @@
...
-        my_data.name,
-        SUM(my_data.x) AS "S"
-    FROM
-        my_data
-    GROUP BY
-        my_data.name
-"""
-)
-expect = pd.DataFrame({"name": ["Alice", "Bob", "Chris"], "S": [14850, 14950, 15050]})
+# This needs to be temprarily disabled since this query requires features that are not yet implemented
```
Comment (ayushdg) [inline continuous_integration/recipe/run_test.py:17]: Out of curiosity what functionality is missing for this to work?

File: dask_planner/src/expression.rs [hunk 7]
```diff
@@ -49,130 +58,69 @@ impl From<ScalarValue> for PyScalarValue {
...
+                                            proj.input.schema().index_of_column(col).unwrap();
+                                        // match &exprs[plan.get_index(col)] {
+                                        match &exprs[col_index] {
+                                            Expr::AggregateFunction { args, .. } => {
+                                                match &args[0] {
+                                                    Expr::Column(col) => {
+                                                        println!(
+                                                            "AGGREGATE COLUMN IS {}",
+                                                            col.name
+                                                        );
```
Comment (ayushdg) [inline dask_planner/src/expression.rs:101]: If the debug prints are useful as work progresses, it might be worth adding a `TODO` comment to clean up later

File: dask_sql/context.py [hunk 8]
```diff
@@ -805,7 +807,11 @@ def _get_ral(self, sql):
                 f"Multiple 'Statements' encountered for SQL {sql}. Please share this with the dev team!"
             )
 
-        nonOptimizedRel = self.context.logical_relational_algebra(sqlTree[0])
+        try:
+            nonOptimizedRel = self.context.logical_relational_algebra(sqlTree[0])
+        except DFParsingException as pe:
+            raise ParsingException(sql, str(pe)) from None
```
Comment (ayushdg) [inline dask_sql/context.py:813]: If one doesn't exist already we should add a test case checking for this.

File: tests/unit/test_mapping.py [hunk 9]
```diff
@@ -3,28 +3,32 @@
...
 
 def test_sql_to_python():
-    assert sql_to_python_value("CHAR(5)", "test 123") == "test 123"
-    assert type(sql_to_python_value("BIGINT", 653)) == np.int64
-    assert sql_to_python_value("BIGINT", 653) == 653
-    assert sql_to_python_value("INTERVAL", 4) == timedelta(milliseconds=4)
+    assert sql_to_python_value(SqlTypeName.VARCHAR, "test 123") == "test 123"
+    assert type(sql_to_python_value(SqlTypeName.BIGINT, 653)) == np.int64
+    assert sql_to_python_value(SqlTypeName.BIGINT, 653) == 653
+    assert sql_to_python_value(SqlTypeName.INTERVAL, 4) == timedelta(microseconds=4000)
```
Comment (ayushdg) [inline tests/unit/test_mapping.py:24]: Is the default datafusion unit in microseconds?

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -162,7 +162,6 @@ jobs:
       - name: Install dependencies and nothing else
         run: |
           conda install setuptools-rust
-          conda install pyarrow>=4.0.0
```
Comment (ayushdg) [inline .github/workflows/test.yml:165]: is the pyarrow requirement no longer needed?

**Review by ayushdg — CHANGES_REQUESTED** [review]
Few minor questions but generally things look good to me

File: continuous_integration/recipe/meta.yaml (hunk 5 above)
Comment (jdye64) [reply continuous_integration/recipe/meta.yaml:33]: No, it is not needed. I had based this PR off the `datafusion-aggregate` one and probably did that before removing this in the other PR. It can be removed and I will remove it.

File: dask_planner/src/expression.rs (hunk 7 above)
Comment (jdye64) [reply dask_planner/src/expression.rs:101]: Yep, my plan is to not merge with upstream master during all of these commits. What that will give us is basically another shot to review before we merge into main as all of these type changes will stick out like red flags and can be changed then but at the same time can be left in for now to assist with current development work.

Long story short. 100% plan on removing them during that final PR

File: dask_sql/context.py (hunk 8 above)
Comment (jdye64) [reply dask_sql/context.py:813]: Several already exist one coming top of mind is the `test_wrong_input` pytest in `test_select.py`

File: tests/unit/test_mapping.py (hunk 9 above)
Comment (jdye64) [reply tests/unit/test_mapping.py:24]: Yes, however that can be changed and I should look into finding out how I can make it default to milliseconds instead

File: .github/workflows/test.yml (hunk 10 above)
Comment (jdye64) [reply .github/workflows/test.yml:165]: Thankfully no. I found a portion of the DataFusion library where those mappings were already present so I just wrote a Rust function that I exposed to Python that allowed me to use that instead of including PyArrow in the Python layer since DataFusion already includes Arrow.

Previously I didn't know that portion of the DataFusion library existed so that was the reason for the original inclusion of PyArrow.

File: continuous_integration/recipe/run_test.py (hunk 6 above)
Comment (jdye64) [reply continuous_integration/recipe/run_test.py:17]: The `SUM()` operation has not yet been implemented as part of the `RexCall -> call.py` port

File: tests/unit/test_mapping.py (hunk 9 above)
Comment (ayushdg) [reply tests/unit/test_mapping.py:24]: Actually, even if it needs a few changes on the python side my personal preference is to staying with this default and evaluating if we need to move to millis if it causes problems down the line

**Review by ayushdg — APPROVED** [review]
Changes lgtm! Thanks a lot @jdye64
